Bump supported versions to 3.9

randyzwitch opened the pull request:
Fixes #2001 

Python 3.9 now publicly available and users starting to mention it on discussion forum. Bump lowest supported version to 3.7, add 3.9, in following the SciPy tradition.

randyzwitch committed:
Bump minor versions to 3.7-3.9

randyzwitch committed:
Update other 3.6 references

randyzwitch committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into add_39_to_circleci

tconkling reviewed (changes requested):
Core chatted about this today and decided that, while we don't "officially" support 3.6 at this point, we'd still like to unofficially support it (and keep it as our minimum version in the code), because it's not currently adding any additional support burden.

(That is: we're not currently maintaining any features that require different codepaths for 3.6 vs 3.7+, and likewise we're not holding back on any features or code patterns that we'd add if only we didn't have to support 3.6.)

So instead, let's keep our minimum version at 3.6, and just bump our maximum to 3.9.

We'll bump the minimum version to 3.7 if and when we have a library dependency that requires it, or there's some other maintenance burden we don't want to deal with.

tconkling commented on .circleci/config.yml:
```diff
@@ -185,11 +185,11 @@ commands:
   circleci:
     jobs:
-      - python-3-6: # Oldest supported Python minor version.
+      - python-3-7: # Oldest supported Python minor version.
```
Can you change this to `python-earliest` - and the other Python job to `python-latest`? (The goal is to not have to modify Github actions every time we bump a Python release, which means having CircleCI job names that stay consistent.)

randyzwitch committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into add_39_to_circleci

randyzwitch committed:
Put back 3.6, switch labels to earliest and latest

randyzwitch committed:
Fix 3.6 image

randyzwitch commented:
@tconkling Unsure if I'm doing it wrong, or our build doesn't actually work under 3.9. Will try to investigate python 3.9 locally.

tconkling commented:
> @tconkling Unsure if I'm doing it wrong, or our build doesn't actually work under 3.9. Will try to investigate python 3.9 locally.

Cool, I'll also take a look on Monday!

randyzwitch commented:
Looks like a Python 3.9 problem. Stops at the same place when trying to install dependencies from Pipfile, 15/39. CircleCI times out at 10 minutes, I'm sure it's been that long on my computer (will let it run for hours to see if it resolves)

Eventually, it barfs out a whole bunch of compiler stuff in relation to pyarrow, so I think we've got a pretty good idea that streamlit doesn't work on 3.9 as-is

randyzwitch commented:
Community might still be catching up, pandas just made 3.9 wheels

https://twitter.com/pandas_dev/status/1321802619527024640

randyzwitch commented:
Good news/bad news...not a Streamlit issue, but a pyarrow pip issue. See the following commands/log that works with a combination of conda and pip:

```
(base) rzwitch@threadripper:~$ conda create -n py39 -c conda-forge python=3.9 -y
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /home/rzwitch/miniconda3/envs/py39

  added / updated specs:
    - python=3.9


The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu
  ca-certificates    conda-forge/linux-64::ca-certificates-2020.6.20-hecda079_0
  certifi            conda-forge/linux-64::certifi-2020.6.20-py39h079e4ff_2
  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.35-h769bd43_9
  libffi             conda-forge/linux-64::libffi-3.2.1-he1b5a44_1007
  libgcc-ng          conda-forge/linux-64::libgcc-ng-9.3.0-h5dbcf3e_17
  libgomp            conda-forge/linux-64::libgomp-9.3.0-h5dbcf3e_17
  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-9.3.0-h2ae2ef3_17
  ncurses            conda-forge/linux-64::ncurses-6.2-he1b5a44_2
  openssl            conda-forge/linux-64::openssl-1.1.1h-h516909a_0
  pip                conda-forge/noarch::pip-20.2.4-py_0
  python             conda-forge/linux-64::python-3.9.0-h2a148a8_4_cpython
  python_abi         conda-forge/linux-64::python_abi-3.9-1_cp39
  readline           conda-forge/linux-64::readline-8.0-he28a2e2_2
  setuptools         conda-forge/linux-64::setuptools-49.6.0-py39h079e4ff_2
  sqlite             conda-forge/linux-64::sqlite-3.33.0-h4cf870e_1
  tk                 conda-forge/linux-64::tk-8.6.10-hed695b0_1
  tzdata             conda-forge/noarch::tzdata-2020d-h516909a_0
  wheel              conda-forge/noarch::wheel-0.35.1-pyh9f0ad1d_0
  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1
  zlib               conda-forge/linux-64::zlib-1.2.11-h516909a_1010


Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate py39
#
# To deactivate an active environment, use
#
#     $ conda deactivate

(base) rzwitch@threadripper:~$ conda activate py39
(py39) rzwitch@threadripper:~$ conda install -c conda-forge pyarrow
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /home/rzwitch/miniconda3/envs/py39

  added / updated specs:
    - pyarrow


The following NEW packages will be INSTALLED:

  abseil-cpp         conda-forge/linux-64::abseil-cpp-20200225.2-he1b5a44_2
  arrow-cpp          conda-forge/linux-64::arrow-cpp-2.0.0-py39h3e27f4d_5_cpu
  aws-c-common       conda-forge/linux-64::aws-c-common-0.4.59-he1b5a44_0
  aws-c-event-stream conda-forge/linux-64::aws-c-event-stream-0.1.6-h84e28f3_5
  aws-checksums      conda-forge/linux-64::aws-checksums-0.1.9-he252421_2
  aws-sdk-cpp        conda-forge/linux-64::aws-sdk-cpp-1.8.70-h9b98462_0
  brotli             conda-forge/linux-64::brotli-1.0.9-he1b5a44_3
  bzip2              conda-forge/linux-64::bzip2-1.0.8-h516909a_3
  c-ares             conda-forge/linux-64::c-ares-1.16.1-h516909a_3
  gflags             conda-forge/linux-64::gflags-2.2.2-he1b5a44_1004
  glog               conda-forge/linux-64::glog-0.4.0-h49b9bf7_3
  grpc-cpp           conda-forge/linux-64::grpc-cpp-1.33.2-h7997a97_0
  krb5               conda-forge/linux-64::krb5-1.17.1-hfafb76e_3
  libblas            conda-forge/linux-64::libblas-3.9.0-2_openblas
  libcblas           conda-forge/linux-64::libcblas-3.9.0-2_openblas
  libcurl            conda-forge/linux-64::libcurl-7.71.1-hcdd3856_8
  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2
  libev              conda-forge/linux-64::libev-4.33-h516909a_1
  libevent           conda-forge/linux-64::libevent-2.1.10-hcdb4288_3
  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-9.3.0-he4bcb1c_17
  libgfortran5       conda-forge/linux-64::libgfortran5-9.3.0-he4bcb1c_17
  liblapack          conda-forge/linux-64::liblapack-3.9.0-2_openblas
  libnghttp2         conda-forge/linux-64::libnghttp2-1.41.0-h8cfc5f6_2
  libopenblas        conda-forge/linux-64::libopenblas-0.3.12-pthreads_h4812303_1
  libprotobuf        conda-forge/linux-64::libprotobuf-3.13.0.1-h8b12597_0
  libssh2            conda-forge/linux-64::libssh2-1.9.0-hab1572f_5
  libthrift          conda-forge/linux-64::libthrift-0.13.0-h5aa387f_6
  libutf8proc        conda-forge/linux-64::libutf8proc-2.5.0-h516909a_2
  lz4-c              conda-forge/linux-64::lz4-c-1.9.2-he1b5a44_3
  numpy              conda-forge/linux-64::numpy-1.19.2-py39h2bb7b6c_1
  orc                conda-forge/linux-64::orc-1.6.5-hd3605a7_0
  parquet-cpp        conda-forge/noarch::parquet-cpp-1.5.1-2
  pyarrow            conda-forge/linux-64::pyarrow-2.0.0-py39h362ff7c_5_cpu
  re2                conda-forge/linux-64::re2-2020.10.01-he1b5a44_0
  snappy             conda-forge/linux-64::snappy-1.1.8-he1b5a44_3
  zstd               conda-forge/linux-64::zstd-1.4.5-h6597ccf_2


Proceed ([y]/n)? y

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
(py39) rzwitch@threadripper:~$ pip install streamlit
Collecting streamlit
  Using cached streamlit-0.70.0-py2.py3-none-any.whl (7.4 MB)
Collecting pillow>=6.2.0
  Using cached Pillow-8.0.1-cp39-cp39-manylinux1_x86_64.whl (2.2 MB)
Collecting click>=7.0
  Using cached click-7.1.2-py2.py3-none-any.whl (82 kB)
Collecting tzlocal
  Using cached tzlocal-2.1-py2.py3-none-any.whl (16 kB)
Processing ./.cache/pip/wheels/bd/1c/8b/7088860a28febba9d6e2dee28e01720ce89c43787a2f7c8cf8/watchdog-0.10.3-py3-none-any.whl
Processing ./.cache/pip/wheels/33/3c/d4/f2d26d7dad1ed59fb6574a88d3026c14f5c48d11748a2e07dd/boto3-1.16.10-py2.py3-none-any.whl
Collecting astor
  Using cached astor-0.8.1-py2.py3-none-any.whl (27 kB)
Collecting toml
  Using cached toml-0.10.2-py2.py3-none-any.whl (16 kB)
Collecting botocore>=1.13.44
  Using cached botocore-1.19.10-py2.py3-none-any.whl (6.7 MB)
Collecting protobuf>=3.6.0
  Using cached protobuf-3.13.0-py2.py3-none-any.whl (438 kB)
Collecting gitpython
  Using cached GitPython-3.1.11-py3-none-any.whl (159 kB)
Collecting altair>=3.2.0
  Using cached altair-4.1.0-py3-none-any.whl (727 kB)
Collecting pydeck>=0.1.dev5
  Using cached pydeck-0.5.0-py2.py3-none-any.whl (4.5 MB)
Collecting base58
  Using cached base58-2.0.1-py3-none-any.whl (4.3 kB)
Collecting enum-compat
  Using cached enum_compat-0.0.3-py3-none-any.whl (1.3 kB)
Collecting cachetools>=4.0
  Using cached cachetools-4.1.1-py3-none-any.whl (10 kB)
Collecting python-dateutil
  Using cached python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)
Processing ./.cache/pip/wheels/50/93/f8/4f0a42a03a06626d675f13907b6982ad5ecff383530af5a900/blinker-1.4-py3-none-any.whl
Collecting pandas>=0.21.0
  Using cached pandas-1.1.4-cp39-cp39-manylinux1_x86_64.whl (9.3 MB)
Collecting tornado>=5.0
  Using cached tornado-6.1-cp39-cp39-manylinux2010_x86_64.whl (427 kB)
Collecting validators
  Using cached validators-0.18.1-py3-none-any.whl (19 kB)
Requirement already satisfied: numpy in ./miniconda3/envs/py39/lib/python3.9/site-packages (from streamlit) (1.19.2)
Requirement already satisfied: pyarrow in ./miniconda3/envs/py39/lib/python3.9/site-packages (from streamlit) (2.0.0)
Collecting packaging
  Using cached packaging-20.4-py2.py3-none-any.whl (37 kB)
Collecting requests
  Using cached requests-2.24.0-py2.py3-none-any.whl (61 kB)
Collecting pytz
  Using cached pytz-2020.4-py2.py3-none-any.whl (509 kB)
Processing ./.cache/pip/wheels/b7/0a/67/ada2a22079218c75a88361c0782855cc72aebc4d18d0289d05/pathtools-0.1.2-py3-none-any.whl
Collecting s3transfer<0.4.0,>=0.3.0
  Using cached s3transfer-0.3.3-py2.py3-none-any.whl (69 kB)
Collecting jmespath<1.0.0,>=0.7.1
  Using cached jmespath-0.10.0-py2.py3-none-any.whl (24 kB)
Collecting urllib3<1.26,>=1.25.4; python_version != "3.4"
  Using cached urllib3-1.25.11-py2.py3-none-any.whl (127 kB)
Requirement already satisfied: setuptools in ./miniconda3/envs/py39/lib/python3.9/site-packages (from protobuf>=3.6.0->streamlit) (49.6.0.post20201009)
Collecting six>=1.9
  Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting gitdb<5,>=4.0.1
  Using cached gitdb-4.0.5-py3-none-any.whl (63 kB)
Collecting entrypoints
  Using cached entrypoints-0.3-py2.py3-none-any.whl (11 kB)
Collecting jinja2
  Using cached Jinja2-2.11.2-py2.py3-none-any.whl (125 kB)
Collecting jsonschema
  Using cached jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
Collecting toolz
  Using cached toolz-0.11.1-py3-none-any.whl (55 kB)
Collecting ipywidgets>=7.0.0
  Using cached ipywidgets-7.5.1-py2.py3-none-any.whl (121 kB)
Collecting traitlets>=4.3.2
  Using cached traitlets-5.0.5-py3-none-any.whl (100 kB)
Collecting ipykernel>=5.1.2; python_version >= "3.4"
  Using cached ipykernel-5.3.4-py3-none-any.whl (120 kB)
Collecting decorator>=3.4.0
  Using cached decorator-4.4.2-py2.py3-none-any.whl (9.2 kB)
Collecting pyparsing>=2.0.2
  Using cached pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
Collecting chardet<4,>=3.0.2
  Using cached chardet-3.0.4-py2.py3-none-any.whl (133 kB)
Requirement already satisfied: certifi>=2017.4.17 in ./miniconda3/envs/py39/lib/python3.9/site-packages (from requests->streamlit) (2020.6.20)
Collecting idna<3,>=2.5
  Using cached idna-2.10-py2.py3-none-any.whl (58 kB)
Collecting smmap<4,>=3.0.1
  Using cached smmap-3.0.4-py2.py3-none-any.whl (25 kB)
Processing ./.cache/pip/wheels/e0/19/6f/6ba857621f50dc08e084312746ed3ebc14211ba30037d5e44e/MarkupSafe-1.1.1-cp39-cp39-linux_x86_64.whl
Collecting attrs>=17.4.0
  Using cached attrs-20.2.0-py2.py3-none-any.whl (48 kB)
Processing ./.cache/pip/wheels/00/cc/12/e712d3979a4aeee423b34a832d97c803e1a67659060315a7e0/pyrsistent-0.17.3-cp39-cp39-linux_x86_64.whl
Collecting widgetsnbextension~=3.5.0
  Using cached widgetsnbextension-3.5.1-py2.py3-none-any.whl (2.2 MB)
Collecting ipython>=4.0.0; python_version >= "3.3"
  Using cached ipython-7.19.0-py3-none-any.whl (784 kB)
Collecting nbformat>=4.2.0
  Using cached nbformat-5.0.8-py3-none-any.whl (172 kB)
Collecting ipython-genutils
  Using cached ipython_genutils-0.2.0-py2.py3-none-any.whl (26 kB)
Collecting jupyter-client
  Using cached jupyter_client-6.1.7-py3-none-any.whl (108 kB)
Collecting notebook>=4.4.1
  Using cached notebook-6.1.4-py3-none-any.whl (9.5 MB)
Collecting pickleshare
  Using cached pickleshare-0.7.5-py2.py3-none-any.whl (6.9 kB)
Collecting prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0
  Using cached prompt_toolkit-3.0.8-py3-none-any.whl (355 kB)
Collecting jedi>=0.10
  Using cached jedi-0.17.2-py2.py3-none-any.whl (1.4 MB)
Collecting pygments
  Using cached Pygments-2.7.2-py3-none-any.whl (948 kB)
Collecting pexpect>4.3; sys_platform != "win32"
  Using cached pexpect-4.8.0-py2.py3-none-any.whl (59 kB)
Collecting backcall
  Using cached backcall-0.2.0-py2.py3-none-any.whl (11 kB)
Collecting jupyter-core
  Using cached jupyter_core-4.6.3-py2.py3-none-any.whl (83 kB)
Collecting pyzmq>=13
  Using cached pyzmq-19.0.2-cp39-cp39-manylinux1_x86_64.whl (1.1 MB)
Collecting prometheus-client
  Using cached prometheus_client-0.8.0-py2.py3-none-any.whl (53 kB)
Collecting Send2Trash
  Using cached Send2Trash-1.5.0-py3-none-any.whl (12 kB)
Collecting terminado>=0.8.3
  Using cached terminado-0.9.1-py3-none-any.whl (13 kB)
Collecting argon2-cffi
  Using cached argon2_cffi-20.1.0-cp35-abi3-manylinux1_x86_64.whl (97 kB)
Collecting nbconvert
  Using cached nbconvert-6.0.7-py3-none-any.whl (552 kB)
Collecting wcwidth
  Using cached wcwidth-0.2.5-py2.py3-none-any.whl (30 kB)
Collecting parso<0.8.0,>=0.7.0
  Using cached parso-0.7.1-py2.py3-none-any.whl (109 kB)
Collecting ptyprocess>=0.5
  Using cached ptyprocess-0.6.0-py2.py3-none-any.whl (39 kB)
Collecting cffi>=1.0.0
  Using cached cffi-1.14.3-cp39-cp39-manylinux1_x86_64.whl (405 kB)
Collecting testpath
  Using cached testpath-0.4.4-py2.py3-none-any.whl (163 kB)
Collecting defusedxml
  Using cached defusedxml-0.6.0-py2.py3-none-any.whl (23 kB)
Collecting nbclient<0.6.0,>=0.5.0
  Using cached nbclient-0.5.1-py3-none-any.whl (65 kB)
Collecting jupyterlab-pygments
  Using cached jupyterlab_pygments-0.1.2-py2.py3-none-any.whl (4.6 kB)
Collecting bleach
  Using cached bleach-3.2.1-py2.py3-none-any.whl (145 kB)
Processing ./.cache/pip/wheels/d7/2c/f8/55fc25b6130566bdddef382806ceaa33161b39a066c2de5912/pandocfilters-1.4.3-py3-none-any.whl
Collecting mistune<2,>=0.8.1
  Using cached mistune-0.8.4-py2.py3-none-any.whl (16 kB)
Collecting pycparser
  Using cached pycparser-2.20-py2.py3-none-any.whl (112 kB)
Collecting async-generator
  Using cached async_generator-1.10-py3-none-any.whl (18 kB)
Collecting nest-asyncio
  Using cached nest_asyncio-1.4.2-py3-none-any.whl (5.3 kB)
Collecting webencodings
  Using cached webencodings-0.5.1-py2.py3-none-any.whl (11 kB)
Installing collected packages: pillow, click, pytz, tzlocal, pathtools, watchdog, urllib3, six, python-dateutil, jmespath, botocore, s3transfer, boto3, astor, toml, protobuf, smmap, gitdb, gitpython, entrypoints, pandas, MarkupSafe, jinja2, attrs, pyrsistent, jsonschema, toolz, altair, prometheus-client, pyzmq, tornado, ipython-genutils, traitlets, jupyter-core, nbformat, pickleshare, wcwidth, prompt-toolkit, decorator, parso, jedi, pygments, ptyprocess, pexpect, backcall, ipython, jupyter-client, ipykernel, Send2Trash, terminado, pycparser, cffi, argon2-cffi, testpath, defusedxml, async-generator, nest-asyncio, nbclient, jupyterlab-pygments, webencodings, pyparsing, packaging, bleach, pandocfilters, mistune, nbconvert, notebook, widgetsnbextension, ipywidgets, pydeck, base58, enum-compat, cachetools, blinker, validators, chardet, idna, requests, streamlit
Successfully installed MarkupSafe-1.1.1 Send2Trash-1.5.0 altair-4.1.0 argon2-cffi-20.1.0 astor-0.8.1 async-generator-1.10 attrs-20.2.0 backcall-0.2.0 base58-2.0.1 bleach-3.2.1 blinker-1.4 boto3-1.16.10 botocore-1.19.10 cachetools-4.1.1 cffi-1.14.3 chardet-3.0.4 click-7.1.2 decorator-4.4.2 defusedxml-0.6.0 entrypoints-0.3 enum-compat-0.0.3 gitdb-4.0.5 gitpython-3.1.11 idna-2.10 ipykernel-5.3.4 ipython-7.19.0 ipython-genutils-0.2.0 ipywidgets-7.5.1 jedi-0.17.2 jinja2-2.11.2 jmespath-0.10.0 jsonschema-3.2.0 jupyter-client-6.1.7 jupyter-core-4.6.3 jupyterlab-pygments-0.1.2 mistune-0.8.4 nbclient-0.5.1 nbconvert-6.0.7 nbformat-5.0.8 nest-asyncio-1.4.2 notebook-6.1.4 packaging-20.4 pandas-1.1.4 pandocfilters-1.4.3 parso-0.7.1 pathtools-0.1.2 pexpect-4.8.0 pickleshare-0.7.5 pillow-8.0.1 prometheus-client-0.8.0 prompt-toolkit-3.0.8 protobuf-3.13.0 ptyprocess-0.6.0 pycparser-2.20 pydeck-0.5.0 pygments-2.7.2 pyparsing-2.4.7 pyrsistent-0.17.3 python-dateutil-2.8.1 pytz-2020.4 pyzmq-19.0.2 requests-2.24.0 s3transfer-0.3.3 six-1.15.0 smmap-3.0.4 streamlit-0.70.0 terminado-0.9.1 testpath-0.4.4 toml-0.10.2 toolz-0.11.1 tornado-6.1 traitlets-5.0.5 tzlocal-2.1 urllib3-1.25.11 validators-0.18.1 watchdog-0.10.3 wcwidth-0.2.5 webencodings-0.5.1 widgetsnbextension-3.5.1
(py39) rzwitch@threadripper:~$ streamlit hello

  Welcome to Streamlit. Check out our demo in your browser.

  Local URL: http://localhost:8501
  Network URL: http://192.168.0.199:8501

  Ready to create your own Python apps super quickly?
  Head over to https://docs.streamlit.io

  May you create awesome apps!

```

randyzwitch committed:
Merge branch 'develop' into add_39_to_circleci

randyzwitch committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into add_39_to_circleci

randyzwitch committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into add_39_to_circleci

randyzwitch committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into add_39_to_circleci

randyzwitch committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into add_39_to_circleci

randyzwitch committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into add_39_to_circleci

randyzwitch committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into add_39_to_circleci

randyzwitch committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into add_39_to_circleci

randyzwitch committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into add_39_to_circleci

randyzwitch committed:
Merge branch 'develop' of https://github.com/streamlit/streamlit into add_39_to_circleci

randyzwitch committed:
Merge branch 'develop' into add_39_to_circleci

randyzwitch commented:
Closing this, since engineering is spending time working on the optimal solution and I'm not participating in that discussion (rightfully!)
